game-app/game-core/src/main/java/games/strategy/triplea/util/TransportUtils.java

```diff
@@ -341,4 +343,46 @@ private static void removeTransportAndLoadedUnits(
   private static Optional<Unit> findEquivalentUnit(final Unit unit, final Collection<Unit> units) {
     return units.stream().filter(unit::isEquivalent).findAny();
   }
+
+  /**
+   * Returns a collection of units with some substituted with equivalent ones to allow unloading.
+   *
+   * <p>If the route isn't an unload or no substitutions were made, the returned collection will
+   * have the same content as the input parameter. Units that can't be unloaded for which no
+   * substitutes could be found will remain in the returned collection.
+   *
+   * @param route The route of the move.
+   * @param units The units being moved.
+   * @return A collection of units with possible substitutions.
+   */
+  public static Collection<Unit> chooseEquivalentUnitsToUnload(
+      final Route route, final Collection<Unit> units) {
+    if (!route.isUnload() || !units.stream().anyMatch(Matches.unitIsLand())) {
+      return new ArrayList<>(units);
+    }
+    final List<Unit> updatedUnits = new ArrayList<>(units);
+    // Loop by index, so we can replace entries within the loop.
+    for (int i = 0; i < updatedUnits.size(); i++) {
+      final Unit origUnit = updatedUnits.get(i);
+      if (!canUnload(origUnit, route.getEnd())) {
+        final int unitIndex = i;
+        // Can't unload the unit - find an equivalent one that can be unloaded
+        // and that's not in our current list of units. Note: This takes into
+        // account previous replacements.
+        route.getStart().getUnits().stream()
+            .filter(u -> u.isEquivalent(origUnit))
+            .filter(Predicate.not(updatedUnits::contains))
+            .filter(u -> canUnload(u, route.getEnd()))
+            // Replace the orig unit with the equivalent one.
+            .findAny()
+            .ifPresent(u -> updatedUnits.set(unitIndex, u));
+      }
+    }
```

Comment:
What do you think about using a stream map function? Something along the lines of:
```java
units.stream()
  .map(origUnit -> {
    if (!canUnload(origUnit, route.getEnd())) {
      return route.getStart().getUnits().stream()
        .filter(u -> u.isEquivalent(origUnit))
        .filter(Predicate.not(units::contains))
        .filter(u -> canUnload(u, route.getEnd()))
        .findAny()
        .orElse(origUnit);
    }
    return origUnit;
  })
```

This avoids the complexity the index introduces, and I believe should behave the same as your current code (note how I changed `updatedUnits` to `units` in the filter)

Reply:
Ah, that's what I was concerned about. Makes sense, but unfortunately it doesn't really make the code simpler to understand :/